src/vm/codepitchingmanager.cpp

```diff
@@ -0,0 +1,508 @@
+    }
+
+    if (s_pExecutedMethods == nullptr)
+    {
+        PtrHashMap *pMap = new (SystemDomain::GetGlobalLoaderAllocator()->GetLowFrequencyHeap()) PtrHashMap();
+        pMap->Init(TRUE, nullptr);
+        s_pExecutedMethods = pMap;
+    }
+
+    UPTR key = (UPTR)GetFullHash(pMD);
+    MethodDesc *pFound = (MethodDesc *)s_pExecutedMethods->LookupValue(key, (LPVOID)pMD);
+    if (pFound == (MethodDesc *)INVALIDENTRY)
+    {
+        s_pExecutedMethods->InsertValue(key, (LPVOID)pMD);
+    }
+
+    return SWA_CONTINUE;
+}
+
+// Visitor for stack walk callback.
+StackWalkAction StackWalkCallback(CrawlFrame* pCf, VOID* data)
+{
+    WRAPPER_NO_CONTRACT;
+
+    // WalkInfo* info = (WalkInfo*) data;
+    return CrawlFrameVisitor(pCf, (Thread *)data);
+}
+
+static ULONGLONG s_PitchedMethodCounter = 0;
+void MethodDesc::PitchNativeCode()
+{
+    WRAPPER_NO_CONTRACT;
+    SUPPORTS_DAC;
+
+    g_IBCLogger.LogMethodDescAccess(this);
+
+    if (!IsPitchable())
+        return;
+
+    PCODE pCode = GetNativeCode();
+
+    if (!pCode)
+        return;
+
+    _ASSERTE(HasPrecode());
+
+    _ASSERTE(HasNativeCode());
+
+    ++s_PitchedMethodCounter;
+
+    if (CLRConfig::GetConfigValue(CLRConfig::INTERNAL_JitPitchMinVal) > s_PitchedMethodCounter)
+    {
+        return;
+    }
+    if (CLRConfig::GetConfigValue(CLRConfig::INTERNAL_JitPitchMaxVal) < s_PitchedMethodCounter)
+    {
+        return;
+    }
+
+    if (LookupOrCreateInNotForPitching(this))
+        return;
+
+    MethodTable * pMT = GetMethodTable();
+    _ASSERTE(pMT != nullptr);
+
+    CodeHeader* pCH = ((CodeHeader*)(pCode & ~1)) - 1;
+    _ASSERTE(pCH->GetMethodDesc() == this);
+
+    HostCodeHeap* pHeap = HostCodeHeap::GetCodeHeap((TADDR)pCode);
+    pHeap->GetJitManager()->FreeCodeMemory(pHeap, (void*)pCode);
+
+    ClearFlagsOnUpdate();
+
+    _ASSERTE(HasPrecode());
+    GetPrecode()->Reset();
+
+    if (HasNativeCodeSlot())
+    {
+        RelativePointer<TADDR> *pRelPtr = (RelativePointer<TADDR> *)GetAddrOfNativeCodeSlot();
+        pRelPtr->SetValueMaybeNull(NULL);
+    }
+    else
+    {
+#ifdef FEATURE_INTERPRETER
+        SetNativeCodeInterlocked(NULL, NULL, FALSE);
+#else
+        SetNativeCodeInterlocked(NULL, NULL);
+#endif
+    }
+
+    _ASSERTE(!HasNativeCode());
+
+    UPTR key = (UPTR)GetFullHash(this);
+    ULONGLONG pitchedBytes;
+    if (CLRConfig::GetConfigValue(CLRConfig::INTERNAL_JitPitchPrintStat) != 0)
+    {
+        SimpleReadLockHolder srlh(s_pPitchingCandidateMethodsLock);
+        pitchedBytes = (ULONGLONG)s_pPitchingCandidateSizes->LookupValue((UPTR)this, nullptr);
+        _ASSERTE(pitchedBytes != (ULONGLONG)INVALIDENTRY);
+        if (pitchedBytes == (ULONGLONG)INVALIDENTRY)
+            pitchedBytes = 0;
+        s_jitPitchedBytes += (pitchedBytes >> 1);
+    }
+    {
+        SimpleWriteLockHolder swlh(s_pPitchingCandidateMethodsLock);
+        s_pPitchingCandidateMethods->DeleteValue(key, (LPVOID)this);
+        s_pPitchingCandidateMethods->Compact();
```

Comment:
@sergign60 Last detail I wonder about - do we need to run the Compact after each DeleteValue? Does the s_pExecutedMethods->Clear() call happen that rarely that we care about the fact that the hash map is not compact?